cpu/qn908x/periph/i2c.c

```diff
@@ -0,0 +1,344 @@
+ *
+ * @}
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "cpu.h"
+#include "mutex.h"
+
+#include "periph_conf.h"
+#include "periph/i2c.h"
+
+#include "vendor/drivers/fsl_clock.h"
+
+#include "gpio_mux.h"
+#include "flexcomm.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static mutex_t locks[I2C_NUMOF];
+
+/**
+ * @brief Limit value I2C CLKDIV register.
+ */
+#define I2C_CLKDIV_MAX 0xffffu
+
+/**
+ * @brief Set the I2C controller mode clock speed.
+ */
+static void _i2c_controller_set_speed(I2C_Type *i2c_dev, uint32_t speed_hz)
+{
+    /* The I2C clock source is based on the FLEXCOMM clock with the following
+     * formula:
+     *   i2c freq := flexcomm freq / ((CLKDIV + 1) * (MSTTIME + 2) * 2)
+     * Where MSTTIME is a number between 0 and 7, and CLKDIV is between 0 and
+     * 0xffff.
+     * The MSTTIME register controls for how many cycles does the clock stay
+     * low and high, allowing to use different values for each one but we only
+     * use symmetric ones here, which is why there's a * 2 in the formula above.
+     */
+    assert(speed_hz > 0);
+    uint32_t bus_freq = CLOCK_GetFreq(kCLOCK_BusClk);
+    uint32_t target = bus_freq / (2 * speed_hz);
+
+    uint32_t best_error = UINT_MAX;
+    uint32_t best_clkdiv = 0;
+    uint32_t best_msttime = 0;
+
+    for (uint32_t msttime_p2 = 9; msttime_p2 >= 2; msttime_p2--) {
+        uint32_t clkdiv_p1 = (target + msttime_p2 / 2) / msttime_p2;
+        if (clkdiv_p1 >= I2C_CLKDIV_MAX + 1) {
+            clkdiv_p1 = I2C_CLKDIV_MAX + 1;
+        }
+        uint32_t error =
+            abs((int32_t)target - (int32_t)(clkdiv_p1 * msttime_p2));
+        if (error < best_error) {
+            best_error = error;
+            best_clkdiv = clkdiv_p1 - 1;
+            best_msttime = msttime_p2 - 2;
+        }
+    }
+    i2c_dev->CLKDIV = I2C_CLKDIV_DIVVAL(best_clkdiv);
+    i2c_dev->MSTTIME = I2C_MSTTIME_MSTSCLLOW(best_msttime) |
+                       I2C_MSTTIME_MSTSCLHIGH(best_msttime);
+    DEBUG("i2c: msttime=%" PRIu32 " clkdiv=%" PRIu32 " error=%" PRIu32 "\n",
+          best_msttime, best_clkdiv, best_error);
+}
+
+static void _i2c_init_pins(i2c_t dev)
+{
+    const i2c_conf_t *conf = &i2c_config[dev];
+
+    /* Configure SDA and SCL pins. SDA/SCL are function 5 on pins up to PA5 and
+     * function 4 on higher pins. */
+    gpio_init_mux(conf->pin_sda, GPIO_T_PIN(conf->pin_sda) <= 5 ? 5 : 4);
+    gpio_init_mux(conf->pin_scl, GPIO_T_PIN(conf->pin_scl) <= 5 ? 5 : 4);
+
+    mutex_unlock(&locks[dev]);
+}
+
+void i2c_init(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+    const i2c_conf_t *conf = &i2c_config[dev];
+    I2C_Type *const i2c_dev = conf->dev;
+
+    int flexcomm_num = flexcomm_init((FLEXCOMM_Type *)i2c_dev, FLEXCOMM_ID_I2C);
+    assert(flexcomm_num >= 0);
+
+    if (flexcomm_num == 1) {
+        /* Disable the FLECOMM1 MULT1/DIV1 divisor. */
+        CLOCK_SetClkDiv(kCLOCK_DivFrg1, 0u);
```

Comment:
Why is FLECOMM1 special?

Reply:
FLEXCOMM0 and 1 support UART and they both have a "fractional" clock divider set for both 0 and 1 in the FC_FRG register. This is different from all the other frequency dividers because instead of taking a clock signal and divide by N which is normally done with a simple counter circuit up to N, this "fractional" one allows you to divide your clock frequency by (1+N/256) where N is between 0 and 255, so you divide by something between 1 and 2. My personal guess is that this is useful in the UART case to run the UART at 9600 bps when your source clock is running at 32 or 32.768 KHz. With only a divisor like you have in the AHB and APB buses your "standard" UART in that case is quite off (8192 or 10922). I think this has some application when going into power down mode but leaving the UART enabled, I haven't tried that.

Anyway, it looks from the manual that this fractional is before the "Flexcomm clock", so this is active even if you use the FLEXCOMM1 for I2C. Your question got me thinking so I hooked up the oscilloscope to check and indeed the I2C clock I see is affected by this. Given the typical frequencies involved in I2C and the fact that it is synchronous so nobody cares about the exact value, I opted for disabling it. This code was setting N to 0, so we divide by (1 + 0/256) which is 1.

FLEXCOMM2 ( == I2C1) doesn't have the fractional divisor.

Reply:
Sounds a lot like Atmel's SERCOMs.

But shouldn't this then be 

```C
if (flexcomm_num == 0 || flexcomm_num == 1) {
```

Reply:
In theory yes, and also use `kCLOCK_DivFrg0` instead of 1 for `flexcomm_num == 0`, but the FLEXCOMM0 doesn't support I2C so we would never be here already.

Reply:
ah turns out those FLEXCOMMs are not as flexible as I thought!
I guess the allocation might change with new members of this chip family, so it would probably good to leave a comment explaining that only FLEXCOM1 and FLEXCOM2 can be I2C.

Reply:
Haha, marketing got you. All the FLEXCOMM blocks have a bit mask at the same offset (0xFF8, PSELID register) that says what functions (I2C, UART, SPI, ...) the given flexcomm supports, and a hardware revision number in case you need to rely on it to know which bugs does the FLEXCOMM block have. I think the idea is more about writing "Flexcomm drivers" that you can share across devices of revisions of the devices and check at runtime that the hardware you are talking to supports your driver, or maybe even discover the capabilities at runtime. For example, the nxp LPC546xx have up to 10 FLEXCOMM and all of them support UART, SPI and I2C, but only two of those support I2S as well. A cursory look at the LPC546xx datasheet shows that the list of registers for those FLEXCOMM are very similar to the QN9080, expect that the QN9080 datasheet shows "reserved" for the bits used for the I2S in that LPC. So there's some potential for reusing these.

That said, the "fractional" divider is not part of the FLEXCOMM block itself, the register this function is changing (`SYSCON->FC_FRG`) is part of the SYSCON block, not the I2C/FLEXCOMM one. In the USART mode I select from `uart.c` the "fractional" divider in SYSCON and the BRG / OSR values in the USARTn/FLEXCOMMn block; so that part is shared. In the LPC546xx example there is apparently only one fractional divider shared among all 10 FLEXCOMM blocks. It kinda makes sense since whether you use the fractional divider depends on whether the AHB is running at 32 KHz or not.

Dividing by 1 is the reset deafult value (MCU reset, not FLEXCOMM reset because the value is in SYSCON). So if you limit this to call `i2c_init()` you are fine even without this call, but if for some reason you are switching between I2C and UART it is possible that the fractional divider was set to something else. What I can do instead is to make `flexcomm_init()` set that value to 0 for the flexcomms that have one (hard-coded to 0 and 1 in this MCU, since this is not really a flexcomm feature), and remove that call from the i2c.c driver. The uart.c driver will set the fractional divider on initialization anyway.